rdrf/rdrf/patients/patient_list_configuration.py

```diff
@@ -34,7 +35,15 @@ def _get_available_columns(self):
         }
 
     def _get_available_facets(self):
-        return ['living_status']
+        def choice_to_option(value, label):
```

Comment:
I guess I went with methods thinking about extensibility after hitting complications with global variables with the report_configuration. However, I can't really think of a scenario where we might hit the same issues here, so I will switch it over for now and we can address any challenges in the future if they occur.

Reply:
Actually.. how could I override these as global variables if I reference the global variable directly in the class methods?

Is it worth declaring them as global variables, but then reference them via their existing getter methods? This would provide an easy way for me to subclass PatientListConfiguration for registry-specific fields/demographics that may need to be displayed in the Patient List.

Reply:
If we want to override it, it can be moved to a class variable like in your example (it will still be 'global' to the class, so be sure not to mutate it, only replace it).

If we want to override it dynamically it could have a getter method, but do we want this class to be overridden?

Reply:
Ok, makes sense. Lets go with your example then.

Reply:
I'm going to modify my implementation & example to work with mutating these values rather than replacing it, which would be a more likely use case.

Reply:
Have a look at the latest commit and let me know if you're OK with how it's been done or if it can be improved.

Reply:
```
class PatientListConfiguration:
    AVAILABLE_COLUMNS = {...}

    def __init__(self, registry):
        self.registry = registry
        self.config = self._load_config()

        self.available_columns = PatientListConfiguration.AVAILABLE_COLUMNS

class ExtendPatientListConfiguration(PatientListConfiguration):
    def __init__(self, registry):
        super().__init__(registry)
        self.available_columns.update({
            'full_name_2': {'label': 'Full Name', 'permission': 'patients.can_see_full_name',
                            'class': ColumnFullName}
        })
```

It's a bit confusing in Python, but what will happen is something like:

1. There's a variable called `AVAILABLE_COLUMNS` on the class `PatientListConfiguration`
2. In the `__init__` method for the class, you make `self.available_columns` on the instance a reference to `AVAILABLE_COLUMNS` (not a copy)
3. When you create a new instance of `ExtendPatientListConfiguration`, you will update its `self.available_columns`, which is a reference to `ExtendPatientListConfiguration.AVAILABLE_COLUMNS`, which is a reference to `PatientListConfiguration.AVAILABLE_COLUMNS`, and so each time you initialise `ExtendPatientListConfiguration`, you will update that column on that instance, on any existing instances, and the base class (and thus any future instances)

Here's a simplified example:

```
In [1]: class Test:
   ...:     VALUE = {"a": 1}
   ...:     def __init__(self):
   ...:         self.value = Test.VALUE
   ...: 

In [2]: Test.VALUE
Out[2]: {'a': 1}

In [3]: t = Test()

In [4]: t.VALUE
Out[4]: {'a': 1}

In [5]: t.value
Out[5]: {'a': 1}

In [6]: class ExtendedTest(Test):
   ...:     def __init__(self):
   ...:         super().__init__()
   ...:         self.value.update({"b": 2})
   ...: 

In [7]: ExtendedTest.VALUE
Out[7]: {'a': 1}

In [8]: t2 = ExtendedTest()

In [9]: t2.value
Out[9]: {'a': 1, 'b': 2}

In [10]: t.value
Out[10]: {'a': 1, 'b': 2}

In [11]: Test.VALUE
Out[11]: {'a': 1, 'b': 2}
```

In this case, because it's an update it won't compound on each subsequent init, if it was an append it would.

So for these classes, two things:

1. I don't think the distinction between `AVAILABLE_COLUMNS` and `self.available_columns` is a useful one. They're a reference to the same thing. Instead, the base class should just define `AVAILABLE_COLUMNS`. If `ExtendPatientListConfiguration` wants to override it, it can do so as it's own class variable, or in the init.
2. Any overriding of class variables that are dicts, lists, or other mutable objects should be done through copying (possibly shallow is enough). Example:

```
In [1]: class Test():
   ...:     VALUE = {"a": 1}
   ...: 

In [2]: class Test2(Test):
   ...:     VALUE = Test.VALUE | {"b": 2}
   ...: 

In [3]: Test.VALUE
Out[3]: {'a': 1}

In [4]: Test2.VALUE
Out[4]: {'a': 1, 'b': 2}
```

Reply:
Updated in the latest commit. Let me know if you're happy and I'll start pushing this along.

Reply:
:smiley:  :+1: